### PR TITLE
RAB-683: Add event listener in order to transfer exported file to storage

### PR DESCRIPTION
**Pull request by StevenVAIDIE**
**Description (for Contributor and Core Developer)**

In this PR: 
- I create the event subscriber in order to launch the transfer to storage
- I also create the handler and the domain

I will implement the real implementation of storage transfer into another PR

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Domain/NoneStorage.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * (c) 2022 Akeneo SAS (https://www.akeneo.com)
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Akeneo\Platform\Bundle\ImportExportBundle\Domain;
+
+class NoneStorage implements StorageInterface
```
Comment (pierrejolly) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Domain/NoneStorage.php:16]:
```suggestion
final class NoneStorage implements StorageInterface
```

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Domain/NoneStorage.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+{
+    public function normalize(): array
+    {
+        return [];
```
Comment (ghoussard) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Domain/NoneStorage.php:20]:
```suggestion
        return [
            'type' => 'none',
        ];
```

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Application/TransferFilesToStorage/FileToTransfer.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+namespace Akeneo\Platform\Bundle\ImportExportBundle\Application\TransferFilesToStorage;
+
+final class FileToTransfer
+{
+    public function __construct(
+        private string $fileKey,
+        private string $storage,
+        private bool $isLocal
```
Comment (ghoussard) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Application/TransferFilesToStorage/FileToTransfer.php:21]: I think we could remove the third paremeter
`isLocal` is computed from the storageAlias in WrittenFile so we can do the same in this model
cf https://github.com/akeneo/pim-community-dev/blob/master/src/Akeneo/Tool/Component/Connector/Writer/File/WrittenFileInfo.php#L62

**Review by ghoussard — APPROVED** [review]

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Application/TransferFilesToStorage/FileToTransfer.php (hunk 3 above)
Comment (StevenVAIDIE) [reply src/Akeneo/Platform/Bundle/ImportExportBundle/Application/TransferFilesToStorage/FileToTransfer.php:21]: I made this class in order to not be coupled with src/Akeneo/Tool/Component/Connector/Writer into the Application layer. If i used this constant we will be coupled again no ?

Comment (ghoussard) [reply src/Akeneo/Platform/Bundle/ImportExportBundle/Application/TransferFilesToStorage/FileToTransfer.php:21]: Yes it's true 👍🏻

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/EventSubscriber/ExportFileToStorageAfterExportSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            return;
+        }
+
+        $jobExecution = $event->getJobExecution();
+        if (JobInstance::TYPE_EXPORT !== $jobExecution->getJobInstance()->getType()) {
+            return;
+        }
+
+        $jobParameters = $jobExecution->getJobInstance()->getRawParameters();
+        if (!array_key_exists('storage', $jobParameters)) {
```
Comment (pierrejolly) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/EventSubscriber/ExportFileToStorageAfterExportSubscriber.php:58]: maybe we can use a constant for "storage"

**Review by pierrejolly — APPROVED** [review]

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Hydrator/LocalStorageHydrator.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    public function hydrate(array $normalizedStorage): StorageInterface
+    {
+        Assert::false($this->versionProvider->isSaaSVersion(), 'Local storage is not available in SaaS version');
+
+        return new LocalStorage($normalizedStorage['file_path']);
+    }
+
+    public function supports(array $normalizedStorage): bool
+    {
+        return array_key_exists('type', $normalizedStorage) && $normalizedStorage['type'] === 'local';
```
Comment (ValentinMumble) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Hydrator/LocalStorageHydrator.php:37]:
```suggestion
        return array_key_exists('type', $normalizedStorage) && $normalizedStorage['type'] === LocalStorage::TYPE;
```

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Hydrator/LocalStorageHydrator.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+final class LocalStorageHydrator implements StorageHydratorInterface
+{
+    public function __construct(private VersionProviderInterface $versionProvider)
+    {
+    }
+
+    public function hydrate(array $normalizedStorage): StorageInterface
+    {
+        Assert::false($this->versionProvider->isSaaSVersion(), 'Local storage is not available in SaaS version');
```
Comment (ValentinMumble) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Hydrator/LocalStorageHydrator.php:30]: We can keep it here but we will need to check before actually exporting no?

**Review by ValentinMumble — APPROVED** [review]

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Hydrator/LocalStorageHydrator.php (hunk 6 above)
Comment (StevenVAIDIE) [reply src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Hydrator/LocalStorageHydrator.php:30]: Yes totally it's just a additionnal check

**Pull request merged**